src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -308,16 +302,71 @@ def validate(self, node, options=None):
                 "supplied node argument should be within an ArrayReference "
                 "node that is within the left-hand-side of an Assignment "
                 "node, but it is on the right-hand-side.")
-        # Does the rhs of the assignment have any operations that
-        # return arrays as this is not currently supported?
+        # Does the rhs of the assignment have any operations that are not
+        # elemental?
         for operation in assignment.rhs.walk(Operation):
-            # At the moment the only array valued operator is matmul
-            if operation.operator == BinaryOperation.Operator.MATMUL:
+            # Allow non elemental UBOUND and LBOUND
+            if operation.operator is BinaryOperation.Operator.LBOUND:
+                continue
+            if operation.operator is BinaryOperation.Operator.UBOUND:
+                continue
+            if not operation.is_elemental():
                 raise TransformationError(
                     "Error in NemoArrayRange2LoopTrans transformation. This "
                     "transformation does not support array valued operations "
                     "on the rhs of the associated Assignment node, but found "
                     "'{0}'.".format(operation.operator.name))
+        # Do not allow to transform expressions with CodeBlocks
+        if assignment.walk(CodeBlock):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. This "
+                "transformation does not support array assignments that "
+                "contain a CodeBlock anywhere in the expression.")
+        # Do not allow to transform expressions with function calls (to allow
+        # this we need to differentiate between elemental and not elemental
+        # functions as they have different semantics in array notation)
+        if assignment.walk(Call):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. This "
+                "transformation does not support array assignments that "
+                "contain a Call anywhere in the expression.")
+
+        for reference in assignment.walk(Reference):
```

Comment:
As a performance optimisation, rather than do repeated walks for each assignment, we could do just one, e.g.
`walk((CodeBlock, Call, Reference))` and then do e.g. `if any(isinstance(node, CodeBlock) for node in nodes):`

Reply:
Done